src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs

```diff
@@ -451,34 +454,25 @@ private SequenceType GetSequenceType()
         private static int GetIndex(int Integer) => Integer & ReadOnlySequence.IndexBitMask;
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private ReadOnlySequence<T> SliceImpl(in SequencePosition start, in SequencePosition end)
+        private ReadOnlySequence<T> SliceImpl(in object? startObject, in int startIndex, in object? endObject, in int endIndex)
```

Comment:
From a quick view of the disassembly (in a similar sample), this pattern seems to help reduce instruction count. So looks good.

[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5ilwDcNek1YAlDgDsMAS3wwWAYQj4ADjIA2fAMp8AbjLAxBNYQGYGsbABMIkjQE8GuDFA5gMDXQEcOMSUYAChC4MrJ2DCAMAJIAor7YGNjAWgA8Pn4BMMGh4ZIAfDQA3jQMZQyqUDJ6iTCWMDZ2jgwQwABWMB4MAPqtHR5C1OUVVTUYdVa29k4y0j2z4wDmfIOl5QD0mwypuBz4+NhQDoVDG1uKVuO4DJIwAO7buDB1YLBcALwARBn+QSFhMjsn3WJ2Gm3W23Wu32h2OazKxAsPyyOQBdgAFH1Op5iMQsR40AwFkTpDBllAAJTwhglU7DMq9drYhjvJh4pkDanDboLMl8FkkpYrakAX1MdIY4MeewOR1BZwhYhgGA4UEk13xnlUh08EC4DAwAAsZNcds8GK8YB9vjBfL9sv88sD5WUpakoTLYS6mBZNQwAOLKgDyHIw6IpLPyPU1qwlbuhsrhca2SpVauuvPJFR1LX1RpNj3NlutyL+uUBkmd1LdHphcupiMFAeV0VJ5PDkfmbeF1GrWx2nvryYhreshlq1zuhuV06gBunFtVsDmZpeby+pYd5aBIKJ11tHGwGgNEAY2EkECN/NXFvXNrtKMdFedLD7EJvxY3B/tqKdu4PR5hE4tjGDcl4MIsh5QOe4x1EaiTzjATiqBAxIYKeV7ONg8gMBoEBgIkFYkoWa5Wl8zDsNwvD8CwSo2EG0ybkUAAqIqfJK3o1gmXoNhYwAQBAx7xIeGi4Oim6/kRl6zhG7xRjy3Zzu8rLSXwLAKUKc4AGRaS0oYsMJR5ifmuDqZqhKqVAZmhhSgxir2tAWC4bhdAAYgJxQ0AAkJU1S1PUjTTHp/SeN0Lg6iGIWDMMvljBMDRTM0frdP41iRdi0XlLF/mTE0MxzGFSRQBgraaZlZTZeMAWJfloWpaVfJQOVDDUrxFp2C4TauRo2CLGwYQALJ4AA1gK5DbKkDBmOQzWNpA6qeMSo4wAg/UYENuCjayAB+3W9WtG3Dc11IANoDTOEDWNEagaOi51Gpd12qBoQaqHkpkAIKLIssC4KEegwK2GizLMiwUgAutSlV1O5ECYqGzhFRguiLPI0iEsS4XFctCAfZI1h7YsuAWQjqUo2jGAY3M9X4yteMEz1RNUhKtL0uUhURQjrJY8jZIU813KpelXSsmTfP+BgAvszzDVZtzSM4/ThMmBKgv47L/Ki+rtO4/jyvNfZwxtbDDADeQ6LC54uDA0Y2hI5bhIcJjNswHbOo44SeGSIsuH+IsRoRqzbNMAA7F4LtPbd1uGK79uhoS6ILBS0e2wrOuEinsecyFCdJ5nbvYzrDAANQMIn0gUlo3sBwbrUSo2JsDaQ4nfo+26SIjOqEpbzguwXGAOwwTtzPnafWCtnt2D7Vf+4agdcuUxBh9oEc3eiPOErcDwSU+GKj9n2K5xX++F+PCAl2XSczwHtminXwxnRdV1r/dhqPTdr3vSwX0/cY/2A/YEG3sIZQ1GP5E2K8Y6R3LsFZkPNB6zCbDLdOxE/RC3jsRYkNMz7M2GEHNmS8bj3AYLDdEC9g7lHgfHchFDO6nxWgwQgZcOaF00gwXShMDojQpGgGhFD0E5z4cHbBDCmHohStrNhHDGZcM2rg2hZRb4SkNuUU6r937PTuk/SOn8KyfW+r9f+QMgFg0hhKaGJCBLhygWvRBO9250MpsRexaIO6pXkWUfB9JCFb0sXDIRbMeYsEDAPUM4ZeGqwUc2EqOt15IwjGIlhMSpEkJkYNbhESon0lSsE4MYSeEBPpCEnG6J3GMOYTTFJnD0lyMKUo4YKiyhqO0S/Fpz1dEdW/gYv+1QAHA0kKDEB5iwFVWYEgJsxTYl2NbmWVxFRd6SFklGFC7dcnJMah2XSONZFHTriKIAA==)

Reply:
> From a quick view of the disassembly (in a similar sample), this pattern seems to help reduce instruction count. 

On Linux, too?

Perf test?

Reply:
@pgovind was going to do some perf validation.

Reply:
Right, I thought this would've performed better too, but it does not and I don't have an explanation for why yet. It seems that calling `SliceImpl(new SequencePosition(someObject, someIndex), someLength)` is somehow faster than calling `SliceImpl(someObject, someIndex, endObject, endLength)` 🤷‍♂️. So, I modified the patch to include just the bug fix and less refactoring to the existing SliceImpl calls. There are only 2 callers to this method now(where the fix is to handle a default(SequencePosition)):
Slice(SequencePosition, long)
Slice(long, SequencePosition)

Reply:
> On Linux, too?

@stephentoub, why do you think the behavior would be different on Linux?

Reply:
> why do you think the behavior would be different on Linux?

Different calling convention?